cfme/common/provider.py

```diff
@@ -165,107 +122,75 @@ def create(self, cancel=False, validate_credentials=True, check_existing=False,
-                        success_text = '{} Providers "{}" was saved'.format(self.string_name,
-                                                                            self.name)
-                        main_view.entities.flash.assert_message(success_text)
-                    else:
-                        add_view.flash.assert_no_error()
-                        raise AssertionError("Provider wasn't added. It seems form isn't accurately"
-                                             " filled")
+        if check_existing and self.exists:
+            created = False
+        else:
+            created = True
 
-            if validate_inventory:
-                self.validate()
+            logger.info('Setting up Infra Provider: %s', self.key)
+            add_view = navigate_to(self, 'Add')
 
-            return created
+            if not cancel or (cancel and any(self.view_value_mapping.values())):
+                # filling main part of dialog
+                add_view.fill(self.view_value_mapping)
 
-        else:
-            # other providers, old code
-            if check_existing and self.exists:
+            if not cancel or (cancel and self.endpoints):
+                # filling endpoints
+                for endpoint_name, endpoint in self.endpoints.items():
+                    try:
+                        # every endpoint class has name like 'default', 'events', etc.
+                        # endpoints view can have multiple tabs, the code below tries
+                        # to find right tab by passing endpoint name to endpoints view
+                        endp_view = getattr(self.endpoints_form(parent=add_view),
+                                            endpoint_name)
+                    except AttributeError:
+                        # tabs are absent in UI when there is only single (default) endpoint
+                        endp_view = self.endpoints_form(parent=add_view)
+
+                    endp_view.fill(endpoint.view_value_mapping)
+
+                    # filling credentials
+                    if hasattr(endpoint, 'credentials'):
+                        endp_view.fill(endpoint.credentials.view_value_mapping)
+                    # sometimes we have cases that we need to validate even though
+                    # there is no credentials, such as Hawkular endpoint
+                    if (validate_credentials and hasattr(endp_view, 'validate') and
+                            endp_view.validate.is_displayed):
+                        # there are some endpoints which don't demand validation like
+                        #  RSA key pair
+                        endp_view.validate.click()
+                        # Flash message widget is in add_view, not in endpoints tab
+                        logger.info(
+                            'Validating credentials flash message for endpoint %s',
+                            endpoint_name)
+                        add_view.flash.assert_no_error()
```

Comment:
Not needed

Reply:
@quarckster, I agree but this is old code which I just shifted when removed main if statement.
I'm planning to refactor whole providers structure soon. So, I'll refactor all code related to common/provider.